mirgecom/diffusion.py

```diff
@@ -105,27 +106,38 @@ class DirichletDiffusionBoundary(DiffusionBoundary):
     .. automethod:: get_diffusion_flux
     """
 
-    def __init__(self, value):
+    def __init__(self, value=None, function=None):
         """
         Initialize the boundary condition.
 
         Parameters
         ----------
-        value: float or meshmode.dof_array.DOFArray
-            the value(s) of $f$ along the boundary
+        value
+            value to prescribe the quantity along the boundary
+
+        function
+            user-defined function prescribing the flux along the boundary
         """
         self.value = value
+        self.function = function
+
+        if value is None and function is None:
+            raise ValueError("Must specify one of 'value' or 'function'")
 
-    def get_grad_flux(self, dcoll, dd_bdry, u_minus):  # noqa: D102
+    def get_grad_flux(self, dcoll, dd_bdry, u_minus, **kwargs):
+        """Flux for gradient evaluation."""
         actx = u_minus.array_context
+        ext_value = self.value if self.function is None else \
+                    self.function(u_minus=u_minus, **kwargs)
```

Comment:
I will need this in the phenolics PR. Decided to put that in this PR to make life easier.

I will need both `time` and `u_minus` in prescribing the flux. Please, can you clarify the idea/purpose of `gradient_numerical_flux_func`?

Reply:
It makes sense for the heat flux BC function to require a boundary value for `u` as input, but I think not so much for the gradient flux (since that's what the function is supposed to determine).

The purpose of a `gradient_numerical_flux_func` would be to customize the numerical flux form used for the gradient (instead of always doing it the way `grad_facial_flux` does). I wouldn't worry about adding something like that right now though; I think defining time/state-dependent BCs in the RHS like we talked about will sidestep a lot of this stuff.